hardware_interface/include/hardware_interface/handle.hpp

```diff
@@ -16,19 +16,126 @@
 #define HARDWARE_INTERFACE__HANDLE_HPP_
 
 #include <string>
+#include <vector>
 #include <utility>
 
 #include "hardware_interface/macros.hpp"
 #include "hardware_interface/visibility_control.h"
 
+#include <hardware_interface/hardware_info.hpp>
+
 namespace hardware_interface
 {
+struct InterfaceConfiguration
+{
+  std::string joint_name;
+  InterfaceInfo interface_info;
+};
+
+/// A variant class used to store value on a given interface
+class Variant
+{
+enum value_type
+{
+  boolean = 0,
+  integer,
+  floating,
+  vector_boolean,
+  vector_integer,
+  vector_floating,
+};
+
+public:
+  Variant(bool value)
+  {
+    value_type_ = value_type::boolean;
+    bool_value_ = value;
+  }
+
+  Variant(int value)
+  {
+    value_type_ = value_type::integer;
+    int_value_ = value;
+  }
+
+  Variant(double value)
+  {
+    value_type_ = value_type::floating;
+    double_value_ = value;
+  }
+
+  Variant(const std::vector<bool> & value)
+  {
+    value_type_ = value_type::vector_boolean;
+    vector_bool_value_ = value;
+  }
+
+  Variant(const std::vector<int> & value)
+  {
+    value_type_ = value_type::vector_integer;
+    vector_int_value_ = value;
+  }
+
+  Variant(const std::vector<double> & value)
+  {
+    value_type_ = value_type::vector_floating;
+    vector_double_value_ = value;
+  }
+
+  void set_value(bool value)
+  {
+    bool_value_ = value;
+  }
+
+  void set_value(int value)
+  {
+    int_value_ = value;
+  }
+
+  void set_value(double value)
+  {
+    double_value_ = value;
+  }
+
+  template <typename DataT, typename std::enable_if<std::is_integral<DataT>, bool> = true>
+  DataT get_value()
+  {
+    return bool_value_;
+  }
+
+  template <typename DataT, typename std::enable_if<std::is_integral<DataT>, int> = true>
+  DataT get_value()
+  {
+    return int_value_;
+  }
+
+  template <typename DataT, typename std::enable_if<std::is_floating_point<DataT>, double> = true>
+  DataT get_value()
+  {
+    return double_value_;
+  }
+
+  bool is_valid()
+  {
+    // Check if nan or empty to return "false"
+  }
+
+private:
+  value_type value_type_;
+  bool bool_value_;
+  int int_value_;
+  double double_value_;
+  std::vector<bool> vector_bool_value_;
+  std::vector<int> vector_int_value_;
+  std::vector<double> vector_double_value_;
+};
+
 /// A handle used to get and set a value on a given interface.
 class ReadOnlyHandle
 {
 public:
   ReadOnlyHandle(
-    const std::string & name, const std::string & interface_name, double * value_ptr = nullptr)
+    const std::string & name, const std::string & interface_name, Variant * value_ptr = nullptr)
```

Comment:
This doesn't fix the fundamental problem I see with the interfaces of ros2_control. With this change, you are still using a pointer to some memory in the hardware_interface for communication between the hardware_interface and the controllers.

Reply:
I just don't know any other approach that would work will all the types we want to support and provides all the flexibility and strictness to manage this sensibly with controller manager.

Using functions does not solve fully all the issues, but adds complexity on the users' side of the code.

Reply:
Using functions makes it explicit and not out of band.